### DISCUSSION TITLE
How to create command with argument, that requires Item (like /give) + how to pass it to another function

### DISCUSSION THREAD

**Discussion by SomeDude-Me** [FabricMC/fabric-api#2337] [Mod Dev Support] [answered]
I found out, that if I do like
`
dispatcher
.register(literal("mycommand")
.then(argument("item", SOMETHING)
.executes(file::function)))
);
`
I am wondering, what is supposed to go in SOMETHING, so the users output would be variable Item (example: integer() makes it int)

+what should i add to the function's brackets to pass thru Item variable? Is it just `Item item`

**Comment by warjort** [comment, marked as answer]
You know if run the ````gensources```` task for your project, you can see the deobfuscated minecraft source?
Then you can look at the give command to see what it does.

To specify the argument you do:
````
argument("item", ItemStackArgumentType.itemStack(registryAccess))
````
see net.minecraft.command.argument.ArgumentTypes

To access the argument from inside your function you do
````
ItemStackArgumentType.getItemStackArgument(context, "item")
````

> +what should i add to the function's brackets to pass thru Item variable? Is it just Item item

Since you are using a function reference you cannot do this, the method signature must match the command functional interface. Instead you get the parameter from the context inside the function as above.

But note the minecraft give command doesn't use function references. Instead it unwraps the parameters at the execute inside a lambda. Your equivalent would be
````
execute(context -> {
   return file.function(context, ItemStackArgumentType.getItemStackArgument(context, "item"));
})
````
the you would add an ItemStack parameter after your context parameter.

**Reply by SomeDude-Me** [reply]
Thanks. It worked... kinda... Inside the function, I have Item variable and I can't do this directly:
`
Item item = ItemStackArgumentType.getItemStackArgument(context, "item");
`
i remember from coding in C# I used `Int.Parse("5");` to make string variable convertible to int variable. Is there equivalent for ItemStackArgument and Item?

**Reply by warjort** [reply]
You mean? https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/user-defined-conversion-operators

java doesn't have implicit conversion operators.

**Reply by SomeDude-Me** [reply]
So there isn't a way to put it in a Item variable?

**Reply by SomeDude-Me** [reply]
> You mean? https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/user-defined-conversion-operators
> 
> java doesn't have implicit conversion operators.

I ment: https://docs.microsoft.com/en-us/dotnet/api/system.int32.parse?view=net-6.0

**Reply by SomeDude-Me** [reply]
`
string s = "5";
int i = int.Parse(s);
i++;
Console.Writeline(i);
//Output: 6
`

**Reply by warjort** [reply]
I don't think I understand your question?
The parameter is an ItemStack (item, count, nbt), the method shown is the way to get it. 
To get an Item you would use getItem() but I would think you want the stack?

**Reply by SomeDude-Me** [reply]
I have object:
`
Item item;
`
When player types /mycommand minecraft:item_frame
i need the minecraft:item_frame to travel thru the method and rewrite item to itsself

(item = item_frame)

and I don't know how to do that


> To get an Item you would use getItem() but I would think you want the stack?

is it `item = ItemStackArgumentType.getItem(context, "item");`

or do i need to change more things? (ItemStackArgumentType.getItem() doesn't exist)

**Reply by warjort** [reply]
````
ItemStack itemStack = ItemStackArgumentType.getItemStackArgument(context, "item");
Item item = itemStack.getItem();
````
The item is just a singleton that defines behaviour of that item.
It is the itemstack that has many instances and you can put in inventories, throw at people, etc.

**Reply by SomeDude-Me** [reply]
> ```
> ItemStack itemStack = ItemStackArgumentType.getItemStackArgument(context, "item");
> Item item = itemStack.getItem();
> ```
> 
> The item is just a singleton that defines behaviour of that item. It is the itemstack that has many instances and you can put in inventories, throw at people, etc.

ItemStack itemStack is an ItemStack type, 
ItemStackArgumentType.getItemStackArgument(context, "item") is an ItemStackType type, therefore it doesn't work
![No name](https://user-images.githubusercontent.com/107791307/174476897-1e13a6e8-e114-4058-abce-9c8975ced0d9.png)


EDIT: it was an easy fix:
`
ItemStackArgument itemStack = ItemStackArgumentType.getItemStackArgument(context, "item");
`

**Reply by warjort** [reply]
My mistake, it should be
````
ItemStackArgument itemStackArgument = ItemStackArgumentType.getItemStackArgument(context, "item");
Item item = itemStackArgument.getItem();
````
I assumed it returned the ItemStack from methods name. It actually returns an ItemStackArgument.

**Reply by SomeDude-Me** [reply]
thanks so mutch. Now i understand it
